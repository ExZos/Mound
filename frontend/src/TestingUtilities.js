import GeneralComponent from './components/GeneralComponent';

class TestingUtilities extends GeneralComponent {
  addUserSessionItem(userID, spaceID, pollID=undefined) {
    const user = {
      id: userID,
      name: 'user' + userID,
      space: spaceID,
      space_name: 'space' + spaceID,
      space_status: true,
      poll: pollID
    };

    this.addToSessionArrayItem('users', user);
  }
}

export default TestingUtilities;
