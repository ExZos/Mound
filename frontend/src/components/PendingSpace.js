import React from 'react';
import { CircularProgress } from '@material-ui/core';

import GeneralComponent from './GeneralComponent';
import { server, api } from '../server';
import '../styles/pendingSpace.scss';

class PendingSpace extends GeneralComponent {
  constructor(props) {
    super(props);

    this.loaded = false;

    this.state = {
      requiredUsers: 3,
      userCount: '',
      remainingUsers: ''
    };
  }

  componentDidMount() {
    this.getUserCountOfSpace();
    this.interval = setInterval(this.getUserCountOfSpace, 1000);
  }

  componentDidUpdate() {
    window.onpopstate = (e) => {
      this.loaded = false;

      this.props.updateState();
      this.getUserCountOfSpace();
    }
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  getUserCountOfSpace = async () => {
    try {
      const res = await server.get(api.getUserCountInSpaceForUser + this.props.user.space + '/' + this.props.user.id);

      this.loaded = true;

      const userCount = res.data['userCount'];
      const user = res.data['user'];

      this.setState({
        userCount: userCount,
        remainingUsers: this.state.requiredUsers - userCount,
        loaded: true
      });

      // Space approved: update user session item
      if(user && this.loaded) {
        this.addToSessionArrayItem('users', user);
        this.props.updateState();
      }
    } catch (e) {
      // TODO: render error component
    }
  }

  displayResultingStatement = () => {
    if(this.state.remainingUsers > 1) {
      return (
        <div>
          {this.state.remainingUsers} more users are required to approve this space.
        </div>
      );
    }
    else if(this.state.remainingUsers === 1) {
      return (
        <div>
          {this.state.remainingUsers} more user is required to approve this space.
        </div>
      );
    }

    return (
      <div>
        Congrats!<br />
        The required user count has been met.<br />
        This space will be approved shortly.
      </div>
    );
  }

  render() {
    if(!this.loaded) {
      return (
        <div id="pendingSpace">
          <CircularProgress color="inherit" />
        </div>
      );
    }

    return(
      <div id="pendingSpace">
        <div className="statusStatement">
          Status: pending
        </div>

        <div className="progressStatement">
          Progress: {this.state.userCount}/{this.state.requiredUsers}
        </div>

        <br />

        <div className="resultingStatement">
          {this.displayResultingStatement()}
        </div>
      </div>
    );
  }
}

export default PendingSpace;
