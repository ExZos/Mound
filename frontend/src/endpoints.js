export const endpoints = {
  messages: '/api/messages/',
  spaces: '/api/spaces/',
  users: '/api/users/',
  poll: '/api/polls/',
  votes: '/api/votes/',
  getSpaceByName: '/space/getByName/',
  getUserCountInSpaceForUser: '/space/getUserCountForUser/',
  getUserByName: '/user/getByName/',
  getUsersInSpace: '/user/getInSpace/',
  getUserInSpaceByName: '/user/getInSpaceByName/',
  getUsersInSpaceExceptName: '/user/getInSpaceExceptName/',
  createUserNApproveSpace: 'user/createNApproveSpace/',
  getMessagesInSpace: '/message/getInSpace/',
  getPendingPollsInSpace: '/poll/getPendingInSpace/',
  getPendingUnvotedPollsForUser: '/poll/getPendingUnvotedForUser/',
  getPendingJoinPollInSpaceByName: 'poll/getPendingJoinInSpaceByName/',
  getJoinPollResults: '/poll/getJoinResults/',
  createNameRelatedPoll: '/poll/createNameRelated/',
  getVotesForPoll: 'vote/getForPoll/',
  getPositiveVotesForPoll: 'vote/getPositivesForPoll/',
  createVoteNUpdatePoll: 'vote/createNUpdatePoll/',
}
