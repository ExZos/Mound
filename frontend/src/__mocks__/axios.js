import { endpoints as api } from '../endpoints';

const axios = {
  create: () => axios,
  get: (url) => {
    switch(url) {
      case api.getSpaceByName + "Headon's Floor/":
        return Promise.resolve({
          data: {
            id: 1,
            name: "Headon's Floor",
            status: true
          }
        });
        break;

      default:
        return Promise.reject();
    }
  }
};

export default axios;
