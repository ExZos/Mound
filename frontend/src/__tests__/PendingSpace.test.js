import React from 'react';
import { shallow, mount, render } from 'enzyme';
import axios from 'axios';
import { createBrowserHistory } from 'history';

import PendingSpace from '../components/PendingSpace';
import TestingUtilities from '../TestingUtilities';
import { endpoints as api } from '../endpoints';

const history = createBrowserHistory();
const tutils = new TestingUtilities();

describe('PendingSpace', () => {
  it('should render approved correctly with props and session', async () => {
    const mockUpdateState = jest.fn();
    const user = {
      id: 1,
      name: 'user1',
      space: 1,
      space_name: 'space1'
    };

    const spy = jest.spyOn(axios, 'get');
    const component = await shallow(<PendingSpace user={user} updateState={mockUpdateState} />);
    const progressStatement = component.find('div.progressStatement');
    const resultingStatement = component.find('div.resultingStatement');

    expect(component).toMatchSnapshot();
    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getUserCountInSpaceForUser + '1/1');
    expect(progressStatement.childAt(1).text()).toEqual('3');
    expect(resultingStatement.text()).toContain('approved');
    expect(mockUpdateState).toHaveBeenCalled();
  });

  it('should render pending correctly with props and session', async () => {
    const mockUpdateState = jest.fn();
    const user = {
      id: 2,
      name: 'user2',
      space: 2,
      space_name: 'space2'
    };

    const spy = jest.spyOn(axios, 'get');
    const component = await shallow(<PendingSpace user={user} updateState={mockUpdateState} />);
    const progressStatement = component.find('div.progressStatement');
    const resultingStatement = component.find('div.resultingStatement');

    expect(component).toMatchSnapshot();
    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getUserCountInSpaceForUser + '2/2');
    expect(progressStatement.childAt(1).text()).toEqual('1');
    expect(resultingStatement.text()).toContain('more users');
    expect(mockUpdateState).not.toHaveBeenCalled();
  });

  it('should render pending correctly with props and session', async () => {
    const mockUpdateState = jest.fn();
    const user = {
      id: 3,
      name: 'user3',
      space: 3,
      space_name: 'space3'
    };

    const spy = jest.spyOn(axios, 'get');
    const component = await shallow(<PendingSpace user={user} updateState={mockUpdateState} />);
    const progressStatement = component.find('div.progressStatement');
    const resultingStatement = component.find('div.resultingStatement');

    expect(component).toMatchSnapshot();
    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getUserCountInSpaceForUser + '3/3');
    expect(progressStatement.childAt(1).text()).toEqual('2');
    expect(resultingStatement.text()).not.toContain('more users');
    expect(resultingStatement.text()).toContain('more user');
    expect(mockUpdateState).not.toHaveBeenCalled();
  });
});
