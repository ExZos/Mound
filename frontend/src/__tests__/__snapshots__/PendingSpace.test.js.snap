// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingSpace should render approved correctly with props and session 1`] = `
<div
  id="pendingSpace"
>
  <div
    className="statusStatement"
  >
    Status: pending
  </div>
  <div
    className="progressStatement"
  >
    Progress: 
    3
    /
    3
  </div>
  <br />
  <div
    className="resultingStatement"
  >
    <div>
      Congrats!
      <br />
      The required user count has been met.
      <br />
      This space will be approved shortly.
    </div>
  </div>
</div>
`;

exports[`PendingSpace should render pending correctly with props and session 1`] = `
<div
  id="pendingSpace"
>
  <div
    className="statusStatement"
  >
    Status: pending
  </div>
  <div
    className="progressStatement"
  >
    Progress: 
    1
    /
    3
  </div>
  <br />
  <div
    className="resultingStatement"
  >
    <div>
      2
       more users are required to approve this space.
    </div>
  </div>
</div>
`;

exports[`PendingSpace should render pending correctly with props and session 2`] = `
<div
  id="pendingSpace"
>
  <div
    className="statusStatement"
  >
    Status: pending
  </div>
  <div
    className="progressStatement"
  >
    Progress: 
    2
    /
    3
  </div>
  <br />
  <div
    className="resultingStatement"
  >
    <div>
      1
       more user is required to approve this space.
    </div>
  </div>
</div>
`;
